cpu/stm32f1/periph/uart.c

```diff
@@ -33,142 +33,109 @@
+int uart_init(uart_t uart, uint32_t baudrate, uart_rx_cb_t rx_cb, void *arg)
 {
-    USART_TypeDef *dev;
-    uint32_t bus_freq;
-    gpio_t rx_pin, tx_pin;
+    uint32_t bus_clk;
     uint16_t mantissa;
     uint8_t fraction;
 
-    /* enable UART and port clocks and select devices */
-    switch (uart) {
-#if UART_0_EN
-        case UART_0:
-            dev = UART_0_DEV;
-            rx_pin = UART_0_RX_PIN;
-            tx_pin = UART_0_TX_PIN;
-            bus_freq = UART_0_BUS_FREQ;
-            /* enable clocks */
-            UART_0_CLKEN();
-            break;
-#endif
-#if UART_1_EN
-        case UART_1:
-            dev = UART_1_DEV;
-            tx_pin = UART_1_TX_PIN;
-            rx_pin = UART_1_RX_PIN;
-            bus_freq = UART_1_BUS_FREQ;
-            /* enable clocks */
-            UART_1_CLKEN();
-            break;
-#endif
-        default:
-            return -2;
+    /* make sure the given device is valid */
+    if (uart >= UART_NUMOF) {
+        return -1;
     }
+
+    /* save ISR context */
+    isr_ctx[uart].rx_cb = rx_cb;
+    isr_ctx[uart].arg   = arg;
+
     /* configure RX and TX pin */
-    gpio_init_af(tx_pin, GPIO_AF_OUT_PP);
-    gpio_init(rx_pin, GPIO_DIR_IN, GPIO_NOPULL);
-
-    /* configure UART to mode 8N1 with given baudrate */
-    bus_freq /= baudrate;
-    mantissa = (uint16_t)(bus_freq / 16);
-    fraction = (uint8_t)(bus_freq - (mantissa * 16));
-    dev->BRR = 0;
-    dev->BRR |= ((mantissa & 0x0fff) << 4) | (0x0f & fraction);
-
-    /* enable receive and transmit mode */
-    dev->CR1 |= USART_CR1_UE | USART_CR1_TE | USART_CR1_RE;
+    gpio_init(uart_config[uart].rx_pin, GPIO_DIR_IN, GPIO_NOPULL);
+    gpio_init_af(uart_config[uart].tx_pin, GPIO_AF_OUT_PP);
+
+    /* enable the clock */
+    clk_en(uart);
+
+    /* reset UART configuration -> defaults to 8N1 mode */
+    dev(uart)->CR1 = 0;
+    dev(uart)->CR2 = 0;
+    dev(uart)->CR3 = 0;
+
+    /* calculate and apply baudrate */
+    bus_clk = (uart_config[uart].bus == APB1) ? CLOCK_APB1 : CLOCK_APB2;
+    bus_clk /= baudrate;
+    mantissa = (uint16_t)(bus_clk / 16);
+    fraction = (uint8_t)(bus_clk - (mantissa * 16));
+    dev(uart)->BRR = ((mantissa & 0x0fff) << 4) | (fraction & 0x0f);
+
+    /* enable the UART's global interrupt and activate it */
+    NVIC_EnableIRQ(uart_config[uart].irqn);
+    dev(uart)->CR1 = (USART_CR1_UE | USART_CR1_TE |
+                      USART_CR1_RE | USART_CR1_RXNEIE);
+
     return 0;
 }
 
 void uart_write(uart_t uart, const uint8_t *data, size_t len)
 {
-    USART_TypeDef *dev;
-
-    switch (uart) {
-#if UART_0_EN
-        case UART_0:
-            dev = UART_0_DEV;
-            break;
-#endif
-#if UART_1_EN
-        case UART_1:
-            dev = UART_1_DEV;
-            break;
-#endif
-    }
-
     for (size_t i = 0; i < len; i++) {
-        while (!(dev->SR & USART_SR_TXE)) {}
-        dev->DR = data[i];
+        while(!(dev(uart)->SR & USART_SR_TXE)) {
+            cpu_sleep_until_event();
```

Comment:
It should not be affected apart from working correctly :)

@lebrush Yes an event is generated because the `TXE` and `UE` interrupts are enabled. :)